lib/service/service.go

```diff
@@ -2339,52 +2358,40 @@ func (process *TeleportProcess) initSSH() error {
 				return trace.Wrap(err)
 			}
 			log.Infof("Service is starting in tunnel mode.")
+			waitFn = agentPool.Wait
+
+			process.OnExit("agentpool.shutdown", func(payload interface{}) {
+				agentPool.Stop()
```

Comment:
Stopping the agentPool will kill all reverse tunnel connections, so this should happen after the server has been waited, while this will happen at the same time as the other shutdowns.

Considering that `*AgentPool`'s `Stop` already internally handles having a nil pointer (by doing nothing) maybe we can just pass the optional `agentPool` pointer into `ssh.shutdown` like it was done before? Moreover, we probably want to handle hard shutdown vs graceful shutdown vs graceful restart differently in the future, and that info is probably only going to be available in the `ssh.shutdown` callback anyway.